python/BioSimSpace/Sandpit/Exscientia/Protocol/_free_energy_mixin.py

```diff
@@ -18,16 +23,16 @@ def __init__(self,
            Parameters
            ----------
 
-           lam : float
+           lam : float or pandas.Series
                The perturbation parameter: [0.0, 1.0]
 
-           lam_vals : [float]
-               The list of lambda parameters.
+           lam_vals : [float] or pandas.DataFrame
+               A list of lambda values.
 
-           min_lam : float
+           min_lam : float or pandas.Series
                The minimum lambda value.
 
-           max_lam : float
+           max_lam : float or pandas.Series
                The maximum lambda value.
```

Comment:
I was just wondering if there was any need to use `pandas.Series` and `pandas.DataFrame`, i.e. could these just be Python dictionaries? In your tests you just pass dictionaries to the constructors, e.g.:

```python
            lam_vals=pd.DataFrame(
                data={'bonded': [0.0, 0.25, 0.5, 0.75, 1.0, 1.0, 1.0],
                      'coul': [0.0, 0.0, 0.0, 0.5, 1.0, 1.0, 1.0],
                      'vdw': [0.0, 0.0, 0.0, 0.0, 0.0, 0.5, 1.0]}))
```

This seems like an additional layer of abstraction that's not needed. (The user might be familiar with Python objects, but not `pandas`.) We currently implicitly depend on `pandas` via other dependencies, so would need to add a hard dependency if used.

If it's convenient to use `pandas` internally, i.e. it makes the validation simpler, e.g. checking all entries have the same number of values, etc., then the data could be stored that way behind the scenes.

(Ultimately I don't mind what you use in the Sandpit, since it is what is convenient for you. Perhaps you are generating these inputs elsewhere, so `Series` and `DataFrame` make sense.)

Reply:
> We currently implicitly depend on pandas via other dependencies, so would need to add a hard dependency if used.

I think in these days, it is very hard to escape pandas in the dependency.

Fo other packages, I think it is fair that we should make the input as general as possible and not rely on specific package, but I think we do expect user to familiar with some basic packages such as numpy, pandas or matplotlib. I'm not quite sure how many users could benefit from using a dictionary instead of a pandas dataframe. 

On the other hand, If we tell the user that we are expecting a pd.dataframe, the user will know what kind of data structure that we are expecting without needing to read the documentation. If we tell user that we are expecting a dictionary, the user will need to read the description to understand how to format the input. 

There is also the benefit that the user could look at the input and have a rough idea of what is happening behind curtain and the benefit of keeping the input and internal data structure same.